### DISCUSSION TITLE
Variable explanations?

### DISCUSSION THREAD

**Discussion by werdnum** [davidusb-geek/emhass#273] [Q&A] [answered]
I was trying to hack in support for heating/cooling, but I'm having a bit of trouble understanding the code.

Would be great to have some explanation for what the variables all mean, especially the more obscure ones.

My first question: what is the difference between P_def_bin1 and P_def_bin2?

**Comment by davidusb-geek** [comment]
Hi sorry for not answering this before.
These are two helper binary variables.
The `P_def_bin1` is used in the LP formulation to solve the problem of a semi-continuous variable: [https://web.mit.edu/lpsolve/doc/semi-cont.htm](https://web.mit.edu/lpsolve/doc/semi-cont.htm)
The `P_def_bin2` variable is used when implementing the constraint that limits the number of startups for each deferrable load.

**Comment by werdnum** [comment]
No problem, do they have the same value? I'm not familiar with the library
you're using (or LP generally), could they be merged into one variable?

— Andrew Garrett

**Reply by davidusb-geek** [reply]
No they don't have the same value and no we need those two variables to be different.

**Comment by werdnum** [comment]
Okay, thanks. What do they represent? I figured one of them is "whether the
load is on or not at timestep T" - what does the other one represent?

— Andrew Garrett

**Comment by davidusb-geek** [comment, marked as answer]
It is an additional binary variable added to be able to solve the problem. These are very typical of the linear programming approach and are basically there to solve for the non-linearities in the equations.
In the specific case of `P_def_bin1` it is used to solve the semi-continuous variable problem.
Please see this link, it is well explained with the corresponding equations: [https://yetanothermathprogrammingconsultant.blogspot.com/2016/08/semi-continuous-variables.html](https://yetanothermathprogrammingconsultant.blogspot.com/2016/08/semi-continuous-variables.html)

**Comment by werdnum** [comment]
Thanks! Just to check my understanding, it sounds like a semi-continuous variable is a variable `x` that can either be zero or between two bounds (they use `l` and `u` in the explanation). My reading is that you model it by adding a binary variable `d`, and then you say that `x` is between `dl` and `du`. If `d` is 0, then `x` has to also be 0, and if `d` is 1 then `x` has to be between `l` and `u`.

That all makes sense. But based on that, it sounds like `d` is 0 when the load is off (planned amount = 0) and 1 when the load is on (planned amount is > 0). That sounds like it's the same? I must be missing something...

I think you've answered my question (that is, that it's there to support modelling the load allocation as a semi-continuous variable) and you don't have to educate me on the ins and outs of linear programming, so I'm happy to take your word for it - just curious...

**Reply by davidusb-geek** [reply]
Yes they are similar. The difference is in the flexibility allowed within the range [l, u]. Even if the load is on (planned amount > 0), the value of x can vary within that specified range. This will add some flexibility during the optimization solve.
